portal-202: Import all users from UnionCloud automatically

tobytwigger opened the pull request.

AidanLaycock committed:
WIP - Command sketch pending UC api Access.

AidanLaycock committed:
Moved twigger/Unioncloud to dev-develop

AidanLaycock committed:
Added getAllUsers Function.

AidanLaycock committed:
Updated twigger/Unioncloud dependancy to dev.

AidanLaycock committed:
Added user Requests rate limit.

AidanLaycock committed:
WIP UC API Users Syncing.

AidanLaycock committed:
Updated sync console command to clean up data processing

Added Delay on first instance when triggered. Updated If logic to ensure next item is triggered.

AidanLaycock committed:
Initial Build of UC data fetcher.

tobytwigger commented on src/Commands/SyncUnionCloudDataUsers.php:
```diff
@@ -0,0 +1,100 @@
+     */
+    public function handle()
+    {
+        $attributes = [
```
This looks very similar to the getUsersData job, could we call that job for the first page syncronously (and pass a parameter to disable triggering the next job)?

AidanLaycock replied:
Yes that should be fine, I'll pick that up tomorrow! :)

tobytwigger commented on src/Jobs/getUsersData.php:
```diff
@@ -0,0 +1,81 @@
+            processUserData::dispatch($User);
+        }
+
+        $this->triggerNext();
```
Will this ever stop? Presumably it should only stop when no data is returned, so we'd need to check that that does stop the job. I have a feeling it throws an exception (possibly here: https://github.com/bristol-su/nus-unioncloud-api-wrapper/blob/master/src/Response/BaseResponse.php#L375), if so you could catch it then stop the job and mark it as a success!

AidanLaycock replied:
It should stop fine as it's checking within the triggerNext function further down in that file that the pagecount is greater than the current page. Which should prevent ever hitting the exception.

tobytwigger replied:
Ah of course, my mistake! :)

tobytwigger commented on src/Jobs/getUsersData.php:
```diff
@@ -0,0 +1,81 @@
+        } catch (\Exception $e) {
+            if($e instanceof ClientException && $e->getCode() === 403) {
+                $this->error('Failed to reach UC');
+                return;
```
Could you also check to see if the rate limit has been reached here? Not sure what the code is, but catching that execption and sending an error message to say the rate limit was met sounds great

AidanLaycock replied:
There doesn't actually seem to be a specific error to do with rate limit reached - https://unioncloudapiv1dev.docs.apiary.io/#introduction/error-codes/1.-generic-request-level-error-codes

Would it just be a 500 error? If so happy to add that in. Would you also like it to retrigger itself but with a 30 seconds wait? Then it should be fine and should get the rest of the data.

tobytwigger replied:
I thought there was an error, but I've just tried to hit it and I'm not getting any errors at all (assuming it isn't documented in the error codes). Maybe just send an entry to the log with the exception message and code, so at least I'm aware of any teething issues with this

tobytwigger commented on src/Jobs/getUsersData.php:
```diff
@@ -0,0 +1,81 @@
+use Illuminate\Queue\SerializesModels;
+use BristolSU\UnionCloud\UnionCloud\UnionCloud;
+
+class getUsersData implements ShouldQueue
```
Change to GetUsersData

tobytwigger commented on src/Jobs/processUserData.php:
```diff
@@ -0,0 +1,48 @@
+use Illuminate\Queue\SerializesModels;
+use BristolSU\Support\User\User;
+
+class processUserData implements ShouldQueue
```
-> ProcessUserData

tobytwigger commented on src/Jobs/processUserData.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        $this->repository = app(\BristolSU\ControlDB\Contracts\Repositories\User::class);
+
+        $User = $this->repository->getById($this->user['uid']);
```
This will throw a `ModelNotFoundException` if the uid isn't found. Surround with a try catch block e.g.
```
try {
$User = $this->repository->getById($this->user['uid']); 
$this->repository->update($this->user['uid'], $User->id());
} catch (ModelNotFoundException $e) {
$this->repository->create($this->user['uid']);
}
```

AidanLaycock replied:
I've added this locally, do you want the following then refactoring:

```
if($User) {
            // If Exists Update:
            $this->repository->update($this->user['uid'], $User->id());
        } else {
            // If doesn't exist then insert:
            $this->repository->create($this->user['uid']);
        }
```

into:

`$this->repository->update($this->user['uid'], $User->id());`

 and within the Try block?

tobytwigger replied:
I think I get what you mean, in which case yes. Should be the same as the example I gave - you try to get the user by ID (taking into account the below comment (`getByDataProviderId`)). If the user doesn't exist, a ModelNotFoundException (the laravel one) is thrown in the repository, caught and the user created. If a user does exist, it just continues in the try block and updates the user.

Also fyi, the update function takes `$user->id()` first (the control id) and the `uid` second (the data provider ID, where in this case the data provider is UnionCloud)

tobytwigger commented on src/Jobs/processUserData.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        $this->repository = app(\BristolSU\ControlDB\Contracts\Repositories\User::class);
+
+        $User = $this->repository->getById($this->user['uid']);
```
This won't work because you're using the UnionCloud `uid` (the data user ID in portal terms) to find a control user by the control id. There's a `getByDataProviderId` function (or similar) which will get a control user by their data provider user (i.e. the unioncloud uid)

tobytwigger replied:
That repository is the right one, as it's the control user repository and you want to get a control user. You're using `$this->repository->getById($controlId)`, but passing in the unioncloud (the data provider) id. You want to get a control user by the data provider ID, for which there's a `$this->repository->getByDataProviderId($this->user['uid'])` function.

This will return a control user still (if they exist), it's just using different criteria. This is needed because the control ID and data provider ID don't have to be the same.

tobytwigger commented on composer.json:
```diff
@@ -12,7 +12,7 @@
         "php": "^7.3",
         "bristol-su/support": "^4.2",
-        "twigger/unioncloud": "^1.0"
+        "twigger/unioncloud": "dev-develop as 1.0.7"
```
Can now be changed to ^1.0.8 thank you!

tobytwigger reviewed (commented):
@AidanLaycock Left a few minor notes (and spotted the problem I think you're having). Looks almost finished though!

AidanLaycock committed:
Minor Updates.

AidanLaycock committed:
DRY refactor of Sync function and getUsersData job.

AidanLaycock committed:
Added Logger for any other exceptions potentially caused by rate limits

AidanLaycock committed:
Refactor to processUserData job.

AidanLaycock commented:
@tobytwigger - This should now be complete. I've run a full data import, can confirm it takes a while. But the script will stop as expected once all data is imported!

Once UC get back to you about the Updated_at column working, i'd just add a unioncloud:users:sync command that just gets any changes in the last 24 hours and then you should be fine.

tobytwigger commented on src/Commands/SyncUnionCloudDataUsers.php:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return void
+     */
+    public function __construct()
```
Laravel will run this on every single request as it registers the commands, so we should limit what's being done here. The main issue will be injecting UnoinCloud.

By typehinting the interface `BristolSU\UnionCloud\UnionCloud\UnionCloudContract` (not the concretion) in the handle method, the UnionCloud class will only be created when being used

tobytwigger commented on src/Commands/SyncUnionCloudDataUsers.php:
```diff
@@ -0,0 +1,94 @@
+     */
+    protected $description = 'Sync users from UnionCloud';
+
+    // Not currently working as Page Limit doesn't work on the Package.
```
Is this something I need to be aware of?

AidanLaycock replied:
We've discussed it before, UC will return the full 500 records as this isn't working on the UC package. I think there is a comment about it on that package. It actually shouldn't be an issue as it just gets the max number of records in 1 go, which seems to be fine.

tobytwigger replied:
Ah of course, thanks!

tobytwigger commented on src/Commands/SyncUnionCloudDataUsers.php:
```diff
@@ -0,0 +1,94 @@
+        // Get Users from UC API as Array:
+        $Users = $response->getRawData();
+
+        getUsersData::dispatch($this->page, $this->pageCount, $factor, $Users);
```
Should be `GetUsersData` as classes have to be PascalCase so composer doesn't complain about non-compliant namespaces

tobytwigger commented on src/Jobs/getUsersData.php:
```diff
@@ -0,0 +1,90 @@
+        $this->requestLimit = config('unioncloud-portal.users_per_minute');
+
+        // Init Repository
+        $this->repository = app(UnionCloud::class);
```
Running in a queue, I don't think this will work? It'll serialize all the properties on the job, try and serialise the repository but fail (I think). Typehinting the repository on the `handle` method should fix it

tobytwigger replied:
Should also be a local variable to avoid job serialization issues

tobytwigger commented on src/Jobs/getUsersData.php:
```diff
@@ -0,0 +1,90 @@
+
+        foreach($this->users as $User)
+        {
+            processUserData::dispatch($User);
```
Caps for classes again - I'll sort the casing of variables as that's more preference, but class names being lowercase will cause problems

AidanLaycock committed:
Resolving Casing

tobytwigger commented on src/Jobs/processUserData.php:
```diff
@@ -0,0 +1,47 @@
+     */
+    public function handle()
+    {
+        $this->repository = app(\BristolSU\ControlDB\Contracts\Repositories\User::class);
```
Can be typehinted in handle to make it more clear. Also shouldn't be a class property as it may cause serialization issues, should be a local var

tobytwigger committed:
Small fixes

tobytwigger committed:
Refactor more of package

tobytwigger committed:
Cache collection not array

tobytwigger committed:
Code review and changes

- Change membership commands to use event pattern
- Tidied Sync command (moved exception handling into UnionCloud class, tidied interface of UnionCloud to make it easier to use (not using an array, so you don't need to remember how the API works).
- New user is both created in control and put in cache with listeners
- Use the models of the unioncloud package rather than an array of attributes

tobytwigger committed:
Remove unused job

tobytwigger committed:
Merge branch 'portal-202/uc-users-updates' into feature/202-import-users-from-UC
